### PR TITLE
sam: adc: Updated SAM ADC driver.

### PR DESCRIPTION
**Pull request by fallrisk**
Updated SAM ADC to work with new ADC API.

Signed-off-by: Justin Watson <jwatson5@gmail.com>

### CODE REVIEW

File: drivers/adc/adc_sam_afec.c [hunk 1]
```diff
@@ -16,143 +20,250 @@
...
+		channels >>= 1;
+		++channel;
+	}
+
+	error = check_buffer_size(sequence, num_active_channels);
+	if (error) {
+		return error;
+	}
+
+	data->channels = sequence->channels;
```
Comment (MaureenHelm) [inline drivers/adc/adc_sam_afec.c]: I think you need to move this into `adc_context_start_sampling` to handle repeated samplings.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Will you please also enable `tests/drivers/adc/adc_api` for a board that uses this driver?

**Comment by fallrisk** [comment]
@MaureenHelm You want me to add a board configuration to the test_adc.c file? The test looks like one would have to connect the ADC pin to ground. Is there a hardware requirement for this test?

**Comment by MaureenHelm** [comment]
> You want me to add a board configuration to the test_adc.c file?

Yes, I'd like to see a board configuration added to test_adc.c, and an appropriate board yaml updated like this:
https://github.com/zephyrproject-rtos/zephyr/blob/871859a07e212a24a32bc9150dcd13a38374b126/boards/arm/frdm_k64f/frdm_k64f.yaml#L8-L10

This will ensure that the driver gets built at least once in CI. We didn't have CI coverage for this driver before the API change, though, so it's ok to post in another PR after RC1, but ideally before the final release.

> The test looks like one would have to connect the ADC pin to ground. 

Why do you think that? The test checks for non-zero ADC sample values.

> Is there a hardware requirement for this test?

Yes, we'd like to be able to show the driver working on hardware.

**Comment by MaureenHelm** [comment]
Thanks @fallrisk, this looks good. I cherry-picked your commit into #7691, but made one additional change in test_adc.c to make it build properly for the sam board.
```diff
diff --git a/tests/drivers/adc/adc_api/src/test_adc.c b/tests/drivers/adc/adc_api/src/test_adc.c
index 1363f66..48316f2 100644
--- a/tests/drivers/adc/adc_api/src/test_adc.c
+++ b/tests/drivers/adc/adc_api/src/test_adc.c
@@ -98,7 +98,7 @@
 
 #define ADC_DEVICE_NAME                CONFIG_ADC_0_NAME
 #define ADC_RESOLUTION         12
-#define ADC_GAIN_1             ADC_GAIN_1
+#define ADC_GAIN               ADC_GAIN_1
 #define ADC_REFERENCE          ADC_REF_EXTERNAL0
 #define ADC_ACQUISITION_TIME   ADC_ACQ_TIME_DEFAULT
 #define ADC_1ST_CHANNEL_ID     0
```

**Pull request closed without merging**